### PR TITLE
added remote cache with fallback

### PR DESCRIPTION
**Pull request by kenotron**
This adds a fallback mechanism to the lage software... due to the limitations so far with backfill, we need to supply azure blob connection strings to the config via a env variable. This can be "better-managed" with a dotenv config read from the lage.config.js. 

(we can further improve this with more changes to backfill).

# Local Config

In Azure, generate a SAS R/O connection string to be shared with the devs. Please this connection string in a BACKFILL_* env.

# Remote Config

In lage.config.js - turn on "writeRemoteCache" to true
Use a account access key connection string as an environment variable

### CODE REVIEW

File: src/types/CacheOptions.ts [hunk 1]
```diff
@@ -3,4 +3,5 @@ import { Config as BackfillCacheOptions } from "backfill-config";
 export type CacheOptions = BackfillCacheOptions & {
   environmentGlob: string[];
   cacheKey: string;
+  fallbackToLocal: boolean;
```
Comment (dzearing) [inline src/types/CacheOptions.ts:6]: shouldn't this be `fallbackToRemote`

File: src/cache/backfill.ts [hunk 2]
```diff
@@ -43,6 +43,25 @@ export async function cacheHash(
...
+  hash: string | null,
+  id: string,
+  cwd: string,
+  cacheOptions: CacheOptions
+) {
+  if (!hash) {
+    return false;
+  }
+
+  const remoteCacheOptions =  getCacheConfig(cwd, cacheOptions);
```
Comment (dzearing) [inline src/cache/backfill.ts:56]: Shouldn't it try backfilling from local first, and then if miss, try remote?

Comment (evancharlton) [reply src/cache/backfill.ts:56]: Could we change the configuration instead and allow the user to specify an ordered list of caches? If they want to do local -> remote or remote -> local, should `lage` care?

Comment (kenotron) [reply src/cache/backfill.ts:56]: This is a really good point - in addition, should we simply download the remote cache into our local cache?

File: src/cache/RemoteFallbackCacheProvider.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        cwd
+      );
+    }
+  }
+
+  async fetch(hash: string) {
+    const hit = await this.localCacheStorageProvider.fetch(hash);
+
+    if (!hit && this.remoteCacheStorageProvider) {
+      return await this.remoteCacheStorageProvider.fetch(hash);
```
Comment (dzearing) [inline src/cache/RemoteFallbackCacheProvider.ts:41]: This is where you get the remote result and funnel it back into local

**Review by dzearing — APPROVED** [review]
awesome!

**Pull request merged**